NXP-16967: Added ZIP import support.

fmunch opened the pull request:
Allows to send a ZIP file containing a CSV file and all required
blobs. It will be unzipped in nuxeo.csv.blobs.folder after
having cleaned the folder. The first CSV file found will then
be used for the import.

Commit:
NXP-16967: added ZIP import support.

Allows to send a ZIP file containing a CSV file and all required
blobs. It will be unzipped in nuxeo.csv.blobs.folder after
having cleaned the folder. The first CSV file found will then
be used for the import.

efge commented:
Thank you. Do you mind creating a question on http://answers.nuxeo.com about the missing feature you have, then create a JIRA NXP ticket by clicking the "Report a Bug" button on the right, and finally reference this NXP ticket at the beginning of your commit message (like "NXP-1234: Added Zip import support")? This may seem a bit of a bother but it helps us better track the work done on the platform.

fmunch commented:
I tried to create a NXP ticket, but the only available project is SUPNXP on my account. Where can I found the "Report a Bug" button? I just found the "Create" button on jira.nuxeo.com.
Actually, @tiry asked me to send this PR, it is based on work I had to do for a Nuxeo project we work on.

efge commented:
To create a ticket you must first create an Answers question on http://answers.nuxeo.com, and you'll have the "Report a Bug" button on the right. Sorry about the process but it helps tracking stuff. Or if you have SUPNXP access just create the SUPNXP and I'll turn it into a NXP.

fmunch commented:
Question: http://answers.nuxeo.com/questions/12016/feature-nuxeo-csv-zip-import
Ticket: https://jira.nuxeo.com/browse/NXP-16967
I also updated the commit message.

efge commented on src/main/java/org/nuxeo/ecm/csv/CSVImporter.java:
```diff
@@ -43,4 +46,18 @@ String launchImport(CoreSession session, String parentPath, File csvFile, String
+    Set<String> getAcceptedTypes();
+
+    /**
+     * Prepares an uploaded file, before the actual import.
```
Please explain in the Javadoc what "preparing" involves, and what has to be done after the prepare phase.

efge commented on src/main/java/org/nuxeo/ecm/csv/CSVImporterImpl.java:
```diff
@@ -18,21 +18,67 @@
+                fileHandlers.put(id, descriptor);
+            } else {
+                fileHandlers.remove(id);
+            }
```
This kind of registration / unregistration code is still quite common in the platform but does not adequately take into account unregistering things in different order than the registration (partial undeploy of a bundle). It's better to use SimpleContributionRegistry (see the many subclasses that use it). I'll accept the pull request without this fix though.

efge commented on src/main/java/org/nuxeo/ecm/csv/CSVImporterImpl.java:
```diff
@@ -18,21 +18,67 @@
+            } else {
+                acceptedTypes.remove(type);
+            }
+        }
```
Please add a log.warn in an `else` case for invalid extensionPoint values.

efge commented on src/main/java/org/nuxeo/ecm/csv/descriptor/AcceptedTypeDescriptor.java:
```diff
@@ -0,0 +1,47 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the GNU Lesser General Public License
+ * (LGPL) version 2.1 which accompanies this distribution, and is available at
+ * http://www.gnu.org/licenses/lgpl.html
```
Please use lgpl-2.1.html in the URL.

efge commented on src/main/java/org/nuxeo/ecm/csv/descriptor/CSVFileHandlerDescriptorComparator.java:
```diff
@@ -0,0 +1,47 @@
+            return fileHandler1 == null ? -1 : 1;
+        }
+
+        return ObjectUtils.compare(fileHandler1.getOrder(), fileHandler2.getOrder());
```
Your comparator should not return 0 if you have the same order but the handlers are different objects. You should then fall back on comparing the objects themselves, to have a deterministic sort for any arbitrary shuffle of different objects (because 0 means that they would be compared as equal therefore not distinguishable by the comparator).

efge commented on src/main/java/org/nuxeo/ecm/csv/filehandler/CSVImportZipHandler.java:
```diff
@@ -0,0 +1,144 @@
+            Enumeration<? extends ZipEntry> entries = zip.entries();
+            while (entries.hasMoreElements()) {
+                ZipEntry entry = entries.nextElement();
+                File file = new File(csvFolder, entry.getName());
```
This is a security hole, as entry.getName() can be manipulated (by manually crafting a special ZIP file) to contain ".." components. You code should have for each entry:

    if (entry.getName().contains("..")) {
        continue;
    }

efge commented on src/main/java/org/nuxeo/ecm/csv/CSVImporter.java:
```diff
@@ -43,4 +46,18 @@ String launchImport(CoreSession session, String parentPath, File csvFile, String
+     * 
+     * @since 7.3
+     */
+    File prepareUploadedFile(UploadedFile uploadedFile) throws ClientException;
```
`org.richfaces.model.UploadedFile` is a high-level UI-related interface that should not appear in low-level services. Please refactor the code a bit (and the accept() method too) to pass just what you need, from what I see the filename, content type and input stream, as distinct arguments.

Commit:
NXP-16967: improvements

Use of SimpleContributionRegistry for file handlers and accepted types,
LGPL 2.1, deterministic sort in CSVFileHandlerDescriptorComparator,
skip ZIP entries containing ".." and removed the use of UploadedFile
in CSVImporter.

fmunch commented:
[SimpleContributionRegistry](https://github.com/nuxeo/nuxeo/blob/master/nuxeo-runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/model/SimpleContributionRegistry.java) is now used for the contributions. It is based on what I saw in [TargetPackageRegistry](https://github.com/nuxeo/nuxeo-target-platforms/blob/master/nuxeo-target-platforms-core/src/main/java/org/nuxeo/targetplatforms/core/service/TargetPackageRegistry.java), but I am not sure I did it right.
